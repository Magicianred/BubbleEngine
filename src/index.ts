import GameEngine from './engine';

const engine: GameEngine = new GameEngine();
const sceneConfig = [
    { type: 'bitmap', properties: { x: 0, y: 0, source: './assets/earth.png' } },
    { type: 'bitmap', properties: { x: 0, y: 0, source: './assets/moon.png' } },
    { type: 'bitmap', properties: { x: 0, y: 0, source: './assets/sun.png' } },
    { type: 'textfield', properties: { x: 0, y: 0, text: '我做的都是动画，这要臣妾怎么办？', color: '#ffffff' } },
    { type: 'rectangle', properties: { x: 500, y: 500, color: '#6429ec', alpha: 0.7 } }
];

let images: Array<string> = [];
for (let config of sceneConfig) {
    if (config['type'] === 'bitmap') {
        images.push(config['properties']['source']);
    }
}
console.log(images);
engine.start(images, sceneConfig);