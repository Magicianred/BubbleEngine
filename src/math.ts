/**
 * 齐次坐标表示的向量
 * 只支持继承 Vector2 和 Vector3
 * 默认表示为一个点
 */
abstract class Vector {
    x: number;
    y: number;
    z: number;
    w: number;
    constructor(x: number, y: number, z: number, w?: number) {
        this.x = x;
        this.y = y;
        this.z = z;
        if (w) {
            this.w = w;
        } else {
            this.w = 1;
        }
    }
}


/**
 * w为1代表一个点，w为0代表向量
 * 默认为点
 */
class Vector3 extends Vector {
    constructor(x: number, y: number, z: number, w?: number) {
        super(x, y, z, w);
    }
}

class Vector2 extends Vector {
    constructor(x: number, y: number, w?: number) {
        super(x, y, 0, w);
    }
}

abstract class Matrix {
    dimension: number;
    matrix: Array<number>;

    constructor(m?: Array<number>) {
        if (m) {
            if (m.length < this.dimension * this.dimension) {
                throw new Error("矩阵大小错误");
            }
            this.matrix = m;
        } else {
            let temp = new Array(this.dimension * this.dimension);
            // 默认返回单位矩阵
            for (let i = 0; i < this.dimension; i++) {
                const pos = i * this.dimension + i;
                temp[pos] = 1;
            }
            this.matrix = temp;
        }
    }
}

/**
 * 三维空间下的矩阵，为四维方阵
 * 用于表示三维空间的仿射变化
 * 采用GL的左手系
 * 
 * +----+----+----+----+
 * | Xx | Xy | Xz |  0 |  <- x axis
 * +----+----+----+----+
 * | Yx | Yy | Yz |  0 |  <- y axis
 * +----+----+----+----+
 * | Zx | Zy | Zz |  0 |  <- z axis
 * +----+----+----+----+
 * | Tx | Ty | Tz |  1 |  <- 相机位置
 * +----+----+----+----+
*/
export class Matrix4 extends Matrix {
    constructor(m4?: Array<number>) {
        super(m4);
        this.dimension = 4;
    }
}

class Matrix3 extends Matrix {
    constructor(m3?: Array<number>) {
        super(m3);
        this.dimension = 3;
    }
}

/**
 * 两个四维矩阵相乘
 * @returns 数组
 */
function m4Mulm4(a: Array<number>, b: Array<number>): Array<number> {
    let out = new Array(16);
    let a00 = a[0],
        a01 = a[1],
        a02 = a[2],
        a03 = a[3];
    let a10 = a[4],
        a11 = a[5],
        a12 = a[6],
        a13 = a[7];
    let a20 = a[8],
        a21 = a[9],
        a22 = a[10],
        a23 = a[11];
    let a30 = a[12],
        a31 = a[13],
        a32 = a[14],
        a33 = a[15];

    // Cache only the current line of the second matrix
    let b0 = b[0],
        b1 = b[1],
        b2 = b[2],
        b3 = b[3];
    out[0] = b0 * a00 + b1 * a10 + b2 * a20 + b3 * a30;
    out[1] = b0 * a01 + b1 * a11 + b2 * a21 + b3 * a31;
    out[2] = b0 * a02 + b1 * a12 + b2 * a22 + b3 * a32;
    out[3] = b0 * a03 + b1 * a13 + b2 * a23 + b3 * a33;

    b0 = b[4];
    b1 = b[5];
    b2 = b[6];
    b3 = b[7];
    out[4] = b0 * a00 + b1 * a10 + b2 * a20 + b3 * a30;
    out[5] = b0 * a01 + b1 * a11 + b2 * a21 + b3 * a31;
    out[6] = b0 * a02 + b1 * a12 + b2 * a22 + b3 * a32;
    out[7] = b0 * a03 + b1 * a13 + b2 * a23 + b3 * a33;

    b0 = b[8];
    b1 = b[9];
    b2 = b[10];
    b3 = b[11];
    out[8] = b0 * a00 + b1 * a10 + b2 * a20 + b3 * a30;
    out[9] = b0 * a01 + b1 * a11 + b2 * a21 + b3 * a31;
    out[10] = b0 * a02 + b1 * a12 + b2 * a22 + b3 * a32;
    out[11] = b0 * a03 + b1 * a13 + b2 * a23 + b3 * a33;

    b0 = b[12];
    b1 = b[13];
    b2 = b[14];
    b3 = b[15];
    out[12] = b0 * a00 + b1 * a10 + b2 * a20 + b3 * a30;
    out[13] = b0 * a01 + b1 * a11 + b2 * a21 + b3 * a31;
    out[14] = b0 * a02 + b1 * a12 + b2 * a22 + b3 * a32;
    out[15] = b0 * a03 + b1 * a13 + b2 * a23 + b3 * a33;

    return out;
}

/**
 * 两个三维矩阵相乘
 * @returns 数组
 */
function m3Mulm3(a: Array<number>, b: Array<number>): Array<number> {
    let out = new Array(9);
    let a00 = a[0],
        a01 = a[1],
        a02 = a[2];
    let a10 = a[3],
        a11 = a[4],
        a12 = a[5];
    let a20 = a[6],
        a21 = a[7],
        a22 = a[8];

    let b00 = b[0],
        b01 = b[1],
        b02 = b[2];
    let b10 = b[3],
        b11 = b[4],
        b12 = b[5];
    let b20 = b[6],
        b21 = b[7],
        b22 = b[8];

    out[0] = b00 * a00 + b01 * a10 + b02 * a20;
    out[1] = b00 * a01 + b01 * a11 + b02 * a21;
    out[2] = b00 * a02 + b01 * a12 + b02 * a22;

    out[3] = b10 * a00 + b11 * a10 + b12 * a20;
    out[4] = b10 * a01 + b11 * a11 + b12 * a21;
    out[5] = b10 * a02 + b11 * a12 + b12 * a22;

    out[6] = b20 * a00 + b21 * a10 + b22 * a20;
    out[7] = b20 * a01 + b21 * a11 + b22 * a21;
    out[8] = b20 * a02 + b21 * a12 + b22 * a22;
    return out;
}

/**
 * 矩阵相乘的统一接口
 * 只考虑三维、四维矩阵乘法
 */
export function mutiply(a: Matrix, b: Matrix) {
    if (a.dimension !== b.dimension) {
        throw new Error("维数错误");
    }
    const dimension = a.dimension;
    if (dimension === 3) {
        return new Matrix3(m3Mulm3(a.matrix, b.matrix));
    } else if (dimension === 4) {
        return new Matrix4(m4Mulm4(a.matrix, b.matrix));
    } else {
        throw new Error("暂不支持该维度的矩阵运算");
    }
}

// 获得三维空间的平移矩阵
function translation(tx: number, ty: number, tz?: number) {
    let z;
    if (tz) {
        z = tz;
    } else {
        z = tz;
    }
    return new Matrix4([
        1, 0, 0, 0,
        0, 1, 0, 0,
        0, 0, 1, 0,
        tx, ty, z, 1
    ]);
}

/**
 * 三维空间的缩放矩阵
 */
function scale3D(scaleX: number, scaleY: number, scaleZ: number) {
    return new Matrix4([
        scaleX, 0, 0, 0,
        0, scaleY, 0, 0,
        0, 0, scaleZ, 0,
        0, 0, 0, 1
    ]);
}

/**
 *  获取绕x轴旋转的旋转矩阵
 */
function rotateX(angle: number) {
    const angleInRad = angle * Math.PI / 180;
    const c = Math.cos(angleInRad);
    const s = Math.sin(angleInRad);
    return new Matrix4([
        1, 0, 0, 0,
        0, c, s, 0,
        0, -s, c, 0,
        0, 0, 0, 1
    ]);
}

/**
 * 获取绕Y轴旋转的旋转矩阵
 */
function rotateY(angle: number) {
    const angleInRad = angle * Math.PI / 180;
    const c = Math.cos(angleInRad);
    const s = Math.sin(angleInRad);
    return new Matrix4([
        c, 0, -s, 0,
        0, 1, 0, 0,
        s, 0, c, 0,
        0, 0, 0, 1
    ]);
}

/**
 * 获取绕Z轴旋转的旋转矩阵
 */
function rotateZ(angle: number) {
    const angleInRad = angle * Math.PI / 180;
    const c = Math.cos(angleInRad);
    const s = Math.sin(angleInRad);
    return new Matrix4([
        c, s, 0, 0,
        -s, c, 0, 0,
        0, 0, 1, 0,
        0, 0, 0, 1
    ]);
}

