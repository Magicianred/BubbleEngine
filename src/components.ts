import { flatArray2D, Vector2 } from "./math";

// 这边逻辑暂时不敢动，不知道如何抽象
export abstract class GameObject {
    x: number = 0;
    y: number = 0;
    alpha: number = 1;

    draw(gl: WebGL2RenderingContext) {
        this.render(gl);
    }

    abstract render(gl: WebGL2RenderingContext): void;
}

export class Rectangle {

    verties: Array<Vector2> = [];
    height: number;
    width: number;

    // 定对角线的两个点
    constructor(width: number = 2, height: number = 2) {
        // 绘制方法: TRIANGLE_STRIP
        this.verties.push(new Vector2(1, 1));
        this.verties.push(new Vector2(-1, 1));
        this.verties.push(new Vector2(1, -1));
        this.verties.push(new Vector2(-1, -1));
    }
}

