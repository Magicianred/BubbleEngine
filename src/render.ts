import { Rectangle } from "./components";
import { flatArray, Matrix4, Vector2, perspProjectionMatrix, baseTranlate } from "./math";

export class Camera {
    static PERSPECTIVE: number = 0;
    static ORTHOGRAPHIC: number = 1;

    type: number;
    fov: number;    // Field of View in Radian
    aspect: number;
    zNear: number;
    zFar: number;
    projectionMatrix: Matrix4;

    constructor(
        type: number = Camera.PERSPECTIVE,
        width: number = 800,
        height: number = 600,
        fov: number = 45 * Math.PI / 180,
        zNear: number = 0.1,
        zFar: number = 100) {
        if (type === Camera.PERSPECTIVE || type === Camera.ORTHOGRAPHIC) {
            this.type = type;
        }
        else throw new Error("相机类型错误！");

        if (type === Camera.PERSPECTIVE) {
            if (zFar < zNear) throw new Error("透视投影错误");
        }

        this.aspect = width / height;
    }

}

export class Renderer {

    private gl: WebGL2RenderingContext;

    // vertex shader program
    private vsSource: string;

    // fragment shader program
    private fsSource: string;

    constructor(gl: WebGL2RenderingContext, vsSource?: string, fsSource?: string) {
        if (!gl) {
            throw new Error("请传入WebGL2的上下文");
        }
        this.gl = gl;
        if (vsSource) {
            this.vsSource = vsSource;
        }
        else {
            this.vsSource = `
                attribute vec4 aVertexPosition;
        
                uniform mat4 uModelViewMatrix;
                uniform mat4 uProjectionMatrix;
        
                void main() {
                    gl_Position = uProjectionMatrix * uModelViewMatrix * aVertexPosition;
                }
            `;
        }
        if (fsSource) {
            this.fsSource = fsSource;
        } else {
            this.fsSource = `
                void main() {
                    gl_FragColor = vec4(1.0, 1.0, 1.0, 1.0);
                }
            `;
        }
    }

    // 创建指定类型的着色器，上传source源码并编译
    private loadShader(type: number, source: string): WebGLShader {
        const shader = this.gl.createShader(type);
        this.gl.shaderSource(shader, source);
        this.gl.compileShader(shader);

        if (!this.gl.getShaderParameter(shader, this.gl.COMPILE_STATUS)) {
            alert('An error occurred compiling the shaders: ' + this.gl.getShaderInfoLog(shader));
            this.gl.deleteShader(shader);
            return null;
        }

        return shader;
    }

    // 创建着色程序
    private initShaderProgram(): WebGLProgram {
        const vertexShader = this.loadShader(this.gl.VERTEX_SHADER, this.vsSource);
        const fragmentShader = this.loadShader(this.gl.FRAGMENT_SHADER, this.fsSource);

        const shaderProgram = this.gl.createProgram();
        this.gl.attachShader(shaderProgram, vertexShader);
        this.gl.attachShader(shaderProgram, fragmentShader);
        this.gl.linkProgram(shaderProgram);

        // 创建失败， alert
        if (!this.gl.getProgramParameter(shaderProgram, this.gl.LINK_STATUS)) {
            alert('Unable to initialize the shader program: ' + this.gl.getProgramInfoLog(shaderProgram));
            return null;
        }

        return shaderProgram;
    }

    // 2D顶点缓冲器
    private initBuffers2D(vertices: Array<Vector2>): WebGLBuffer {
        console.log(new Float32Array(flatArray(vertices)));
        const positionBuffer = this.gl.createBuffer();
        this.gl.bindBuffer(this.gl.ARRAY_BUFFER, positionBuffer);
        this.gl.bufferData(this.gl.ARRAY_BUFFER, new Float32Array(flatArray(vertices)), this.gl.STATIC_DRAW);
        return {
            position: positionBuffer
        }
    }

    private drawScene(camera: Camera, programInfo: any, buffers: WebGLBuffer) {
        this.gl.clearColor(0, 0, 0, 1);
        this.gl.clearDepth(1); // clear everything
        this.gl.enable(this.gl.DEPTH_TEST);
        this.gl.depthFunc(this.gl.LEQUAL);

        this.gl.clear(this.gl.COLOR_BUFFER_BIT | this.gl.DEPTH_BUFFER_BIT);


        // 正交投影先不考虑，先只考虑了透视投影
        let projection = perspProjectionMatrix(
            camera.fov,
            camera.aspect,
            camera.zNear,
            camera.zFar
        );

        let modelViewMatrix = new Matrix4();
        modelViewMatrix = baseTranlate([0, 0, -6], modelViewMatrix);

        {
            const numComponents = 2;
            const type = this.gl.FLOAT;
            const normalize = false;
            const stride = 0;
            const offset = 0;
            this.gl.bindBuffer(this.gl.ARRAY_BUFFER, buffers['position']);
            this.gl.vertexAttribPointer(
                programInfo.attribLocations.vertexPosition,
                numComponents,
                type,
                normalize,
                stride,
                offset
            );
            this.gl.enableVertexAttribArray(
                programInfo.attribLocations.vertexPosition);
        }


        this.gl.useProgram(programInfo.program);

        // set shader uniforms
        this.gl.uniformMatrix4fv(
            programInfo.uniformLocations.projectionMatrix,
            false,
            projection.matrix);
        this.gl.uniformMatrix4fv(
            programInfo.uniformLocations.modelViewMatrix,
            false,
            modelViewMatrix.matrix);

        {
            const offset = 0;
            const vertexCount = 4;
            this.gl.drawArrays(this.gl.TRIANGLE_STRIP, offset, vertexCount);
        }
    }

    public render(camera: Camera) {
        const shaderProgram = this.initShaderProgram();

        const programInfo = {
            program: shaderProgram,
            attribLocations: {
                vertexPosition: this.gl.getAttribLocation(shaderProgram, 'aVertexPosition'),
            },
            uniformLocations: {
                projectionMatrix: this.gl.getUniformLocation(shaderProgram, 'uProjectionMatrix'),
                modelViewMatrix: this.gl.getUniformLocation(shaderProgram, 'uModelViewMatrix'),
            },
        };

        let a = new Vector2(-1, 1);
        let c = new Vector2(1, -1);
        let square = new Rectangle(a, c);
        const buffers = this.initBuffers2D(square.verties);

        this.drawScene(camera, programInfo, buffers);
    }
}