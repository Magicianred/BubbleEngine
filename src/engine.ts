import { Circle, GameObject, Rectangle, Triangle } from './entities';
import { Camera, Renderer } from './render';
const canvas = document.getElementById('game') as HTMLCanvasElement;
const gl = canvas.getContext('webgl2');

const magicNumber = {
    textOffset: 20 /** 文字的偏移量，改变锚点 */
}

const imagesStorage: any = {};

function getImage(url: string): ImageBitmap {
    return imagesStorage[url];
}

function loadImage(url: string, onLoad: Function): void {
    const image = new Image();
    image.src = url;
    image.onload = function () {
        imagesStorage[url] = image;
        onLoad(image);
    };
}

function loadMultiImages(urls: string[], onLoad: Function): void {
    if (urls.length < 1) {
        onLoad();
        return;
    }
    let count = 0;
    for (const url of urls) {
        loadImage(url, image => {
            count++;
            imagesStorage[url] = image;
            if (count === urls.length) {
                onLoad();
            }
        });
    }
}

export class GameEngine {

    private fps: number = 60;
    private mileSecondPerFrame: number = 1000 / this.fps;
    private iterationCurrentTime: number = 0;


    public onTick?: () => void;
    public onUpdate?: (deltaTime: number) => void;
    public onStart?: () => void;

    private lastTime: number = 0;

    private renderList: GameObject[] = [];
    private gameObjects: { [id: string]: GameObject } = {};

    private renderer: Renderer;
    private camera: Camera;

    public addGameObject(gameObject: GameObject): void {
        this.renderList.push(gameObject);
    }

    public getGameObjecet(id: string): GameObject {
        return this.gameObjects[id];
    }

    createGameObject(type: string, properties: any): GameObject {
        let gameObject: GameObject;
        switch (type) {
            case 'rectangle':
                gameObject = new Rectangle(properties.width, properties.height, properties.color);
                break;
            case 'circle':
                gameObject = new Circle(properties.radius, properties.color);
                break;
            case 'triangle':
                gameObject = new Triangle(properties.color);
                break;
            default:
                throw new Error("不存在该物件");
        }
        return gameObject;
    }

    private enterFrame(advancedTime: number) {
        const deltaTime = advancedTime - this.lastTime;
        this.lastTime = advancedTime;
        this.onEnterFrame(deltaTime);
        requestAnimationFrame(advancedTime => this.enterFrame(advancedTime));
    }

    private loadScene(sceneConfig: any): void {
        this.camera = new Camera();
        for (let config of sceneConfig) {
            const object = this.createGameObject(config.type, config.properties);
            if (config.id) {
                this.gameObjects[config.id] = object;
            }
            this.addGameObject(object);
        }
        this.renderer = new Renderer(gl, this.renderList);
    }

    private onEnterFrame(deltaTime: number) {

        this.iterationCurrentTime += deltaTime;
        while (this.iterationCurrentTime >= this.mileSecondPerFrame) {
            this.iterationCurrentTime -= this.mileSecondPerFrame;
            if (this.onTick) {
                this.onTick();
            }
        }
        if (this.onUpdate) {
            this.onUpdate(deltaTime);
        }
        this.renderer.render(this.camera);
    }

    public start(images: string[], sceneConfig: any, onComplete?: Function): void {
        const initialTime = Date.now();
        loadMultiImages(images, () => {
            this.loadScene(sceneConfig);
            if (this.onStart) {
                this.onStart();
            }
            const advancedTime = Date.now() - initialTime;
            this.enterFrame(advancedTime);
            if (onComplete) {
                onComplete();
            }
        });

    }
}