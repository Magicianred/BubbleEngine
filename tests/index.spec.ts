import { GameEngine, TextField } from '../src/engine';

let canvas: HTMLCanvasElement;
describe('GameEngine', () => {
    beforeEach(() => {
        canvas = document.createElement('canvas');
        canvas.id = 'gameCanvas';
        canvas.width = 800;
        canvas.height = 800;
        if (document.body.children[0]) {
            document.body.removeChild(document.body.children[0]);
        }
        document.body.appendChild(canvas);
    });
    describe('TextField', () => {
        it('render', () => {
            const gameEngine = new GameEngine();
            const textField = new TextField();
            textField.text = 'Hello,World';
            gameEngine.addGameObject(textField);
            const context = canvas.getContext('2d') as any;
            gameEngine.draw(context);

            const dc = context.__getDrawCalls();
            expect(dc).toMatchSnapshot();
        })
        it('render-2', () => {
            const gameEngine = new GameEngine();
            const textField = new TextField();
            textField.text = 'Hello, World';
            gameEngine.addGameObject(textField);
            const context = canvas.getContext('2d') as any;
            gameEngine.draw(context);
            const dc = context.__getDrawCalls();
            expect(dc).toMatchSnapshot();
        })
    });
});
