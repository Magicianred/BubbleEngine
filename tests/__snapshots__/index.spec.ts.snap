// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameEngine TextField render 1`] = `
Array [
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "Hello,World",
      "x": 0,
      "y": 20,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
]
`;

exports[`GameEngine TextField render-1 1`] = `
Array [
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "Hello, World",
      "x": 0,
      "y": 20,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
]
`;

exports[`GameEngine TextField render-2 1`] = `
Array [
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "Hello, World",
      "x": 0,
      "y": 20,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
]
`;
